### ISSUE TITLE
[CNNScoreVariants] ValueError('Error! Unknown code:', '\x00')

### ISSUE THREAD

**Issue by heliac2000** [broadinstitute/gatk#4727]
Environment:

*   GATK 4.0.3.0
*   Python 2.7.14
*   Red Hat Enterprise Linux Server release 7.4 (Maipo)

I'm using GATK CNNScoreVariants 1D model(CPU mode) for hundreds of VCF files,  but the following error occurred in some of them.

    00:13:26.470 INFO  ProgressMeter - Traversal complete. Processed 4924627 total variants in 79.6 minutes.
    00:14:14.422 INFO  CNNScoreVariants - Shutting down engine
    [2018/05/02 13:10:44 JST] org.broadinstitute.hellbender.tools.walkers.vqsr.CNNScoreVariants done. Elapsed time: 79.75 minutes.
    Runtime.totalMemory()=1126694912
    org.broadinstitute.hellbender.utils.python.PythonScriptExecutorException: Traceback detected: Traceback (most recent call last):
      File "<stdin>", line1, in <module>
      File ".../lib/python2.7/site-packages/vqsr_cnn/inference.py", line 51, in score_and_write_batch
        reference_batch.append(reference_string_to_tensor(fifo_data[4]))
      File ".../lib/python2.7/site-packages/vqsr_cnn/inference.py", line 107, in reference_string_to_tensor
        raise ValueError('Error! Unknown code:', b)
    ValueError('Error! Unknown code:', '\x00')
    >>>
            at org.broadinstitute.hellbender.utils.python.StreamingPythonScriptExecutor.getAccumulatedOutput(StreamingPythonScriptExecutor.java:214)
            at org.broadinstitute.hellbender.tools.walkers.vqsr.CNNScoreVariants.onTraversalSuccess(CNNScoreVariants.java:385)
            at org.broadinstitute.hellbender.engine.GATKTool.doWork(GATKTool.java:894)
            at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:134)
            at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:179)
            at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:198)
            at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:160)
            at org.broadinstitute.hellbender.Main.mainEntry(Main.java:203)
            at org.broadinstitute.hellbender.Main.main(Main.java:289)
    
I tried again, but same message are shown.

**Comment by heliac2000** [comment]
Thanks for your quick response. I will try again with Python3 next week.

**Comment by zzxzxzzxz** [comment]
Hi. I encounter the same error with GATK4.0.4.0 and the python environment created by gatkcondaenv.yml

```
14:49:35.361 INFO  CNNScoreVariants - Using key:CNN_1D for CNN architecture:/tmp/zzxzxzzxz/1d_cnn_mix_train_full_bn.4552731615279398677.json and weights:/tmp/zzxzxzzxz/1d_cnn_mix_train_full_bn.2635334538442041575.hd5
14:49:36.747 INFO  ProgressMeter - Starting traversal
14:49:36.747 INFO  ProgressMeter -        Current Locus  Elapsed Minutes    Variants Processed  Variants/Minute
14:49:36.768 INFO  ProgressMeter -             unmapped              0.0                     1           3157.9
14:49:36.769 INFO  ProgressMeter - Traversal complete. Processed 1 total variants in 0.0 minutes.
14:49:36.772 INFO  CNNScoreVariants - Shutting down engine
[May 9, 2018 2:49:36 PM CST] org.broadinstitute.hellbender.tools.walkers.vqsr.CNNScoreVariants done. Elapsed time: 0.08 minutes.
Runtime.totalMemory()=41160278016
org.broadinstitute.hellbender.utils.python.PythonScriptExecutorException: Traceback detected: Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/zzxzxzzxz/anaconda3/envs/gatk/lib/python3.6/site-packages/vqsr_cnn/vqsr_cnn/inference.py", line 51, in score_and_write_batch
    reference_batch.append(reference_string_to_tensor(fifo_data[4]))
  File "/home/zzxzxzzxz/anaconda3/envs/gatk/lib/python3.6/site-packages/vqsr_cnn/vqsr_cnn/inference.py", line 107, in reference_string_to_tensor
    raise ValueError('Error! Unknown code:', b)
ValueError: ('Error! Unknown code:', '\x00')
>>>
	at org.broadinstitute.hellbender.utils.python.StreamingPythonScriptExecutor.getAccumulatedOutput(StreamingPythonScriptExecutor.java:214)
	at org.broadinstitute.hellbender.tools.walkers.vqsr.CNNScoreVariants.onTraversalSuccess(CNNScoreVariants.java:390)
	at org.broadinstitute.hellbender.engine.GATKTool.doWork(GATKTool.java:894)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:134)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:179)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:198)
	at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:160)
	at org.broadinstitute.hellbender.Main.mainEntry(Main.java:203)
	at org.broadinstitute.hellbender.Main.main(Main.java:289)
```

These are the vcf files I used.
http://w.csie.org/~r03944026/vcf/ok.vcf
http://w.csie.org/~r03944026/vcf/error.vcf

ok.vcf is able to be processed successfully, while error.vcf is not.
The only different between these two file is the position.

**Comment by lucidtronix** [comment]
I see that this is occurring in the mitochondrial chromosome.  The model was trained mostly with the autosomes and so scores on the mitochondrial DNA have not been extensively validated.  That said, this looks like a bug and we hope to have a fix in soon.

**Comment by lucidtronix** [comment]
& Thanks for the example VCFs that is a big help!

**Comment by heliac2000** [comment]
Thanks zzxzxzzxz. 

I insert a debug print to `score_and_write_batch()` function in `inference.py` and run CNNScoreVariants with `err.vcf`.
The following data is received by python process via fifo:

    'chrM\t16521\tC\t[T]\tGGGCCCATAACACTTGGGGGTAGCTAAAGTGAACTGTATCCGACATCTGGTTCCTACTTCAGGGCCATAAAGCCTAAATAGCCCACACGTTCCCCTTAAATAAGACATCACGATG\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\tAC=2;MQRankSum=-0.712;MQ=57.64;AF=1.00;MLEAC=2;BaseQRankSum=5.566;ExcessHet=3.0103;MLEAF=1.00;DP=4492;ReadPosRankSum=-1.183;AN=2;FS=0.000;QD=30.32;SOR=0.761;ClippingRankSum=0.053;\tSNP\n'

There are 13 NULL data in reference bases. This data are created by `transferToPythonViaFifo()` method in `CNNScoreVariants.java`.

    private void transferToPythonViaFifo(final VariantContext variant, final ReferenceContext referenceContext) {
        try {
            final String outDat = String.format("%s\t%s\t%s\t%s\n",
                    getVariantDataString(variant),
             --->   new String(Arrays.copyOfRange(referenceContext.getBases(), 0, windowSize), "UTF-8"),
                    getVariantInfoString(variant),
                    variant.isSNP() ? "SNP" : variant.isIndel() ? "INDEL" : "OTHER");

    (windowSize == 128) / 2 - (chrM length(16571) - variant position(16521) + 1) = 13

I add the `fifo_data[4] = fifo_data[4].strip('\x00')` in `score_and_write_batch()` function and rerun:

    ValueError: Error when checking : expected reference to have shape (128, 4) but got array with shape (115, 4)

What should I do ?

**Comment by cmnbroad** [comment]
@heliac2000 There is a PR (https://github.com/broadinstitute/gatk/pull/4751) in progress with a fix for this issue. Feel free to try that branch, which will probably solve your immediate problem.

**Comment by heliac2000** [comment]
@cmnbroad Thanks! Very thanks!

**Comment by cmnbroad** [comment]
Fixed by #4751.
